File: src/main/scala/a_lists/P03.scala
package a_lists

import java.util.NoSuchElementException
import java.lang.IndexOutOfBoundsException


object P03 {
  
  // P03. Find the Kth element of a list. (The first element is the 0th)
  def kthRecursive[A](k: Int, xs: List[A]): A = (k, xs) match {
    case (0, h :: _ )   => h
    case (k, _ :: tail) => kth(k - 1, tail)
    case _              => throw new NoSuchElementException
  }

  def kth[A](k: Int, xs: List[A]): A = {
    if(xs.length <= k) {
      throw new IndexOutOfBoundsException
    }
    xs(k)
  }
}

File: ruby/a_lists/p03.rb
class P03

  # Find the Kth element in a list (the first element is the "0th")

  def initialize(list)
    @list = list
  end

  def kth_element(k = 0)
    @list[k] if @list
  end

  def find_by_index(k = 0)
    @list&.each_with_index do |elem,i|
      return elem if i == k
    end
    nil
  end

end